File: Medium/C++/53.maximum-subarray.cpp
/*
 * @lc app=leetcode id=53 lang=cpp
 *
 * [53] Maximum Subarray
 */
#include <iostream>
#include <vector>

using namespace std;

void printIntVector(vector<int> &nums)
{
    for (auto itr = nums.begin(); itr != nums.end(); itr++)
    {
        cout << "Vector[" << itr - nums.begin() << "] : " << *itr << endl;
    }
}
// @lc code=start
class Solution
{
public:
    int maxSubArray(vector<int> &nums)
    {
        int maxSum = nums[0];
        int sum = nums[0];
        for (int i = 1; i < nums.size(); i++)
        {
            sum = max(nums[i] + sum, nums[i]);
            if (sum > maxSum)
            {
                maxSum = sum;
            }
        }
        return maxSum;
    }
};
// @lc code=end

int main()
{
    Solution sol;
    vector<int> nums{-2, 1, -3, 4, -1, 2, 1, -5, 4};
    printIntVector(nums);
    int res = sol.maxSubArray(nums);
    cout << res << endl;
    return 0;
}

File: Medium/Python/53.maximum-subarray.py
#
# @lc app=leetcode id=53 lang=python3
#
# [53] Maximum Subarray
#
from typing import List

# @lc code=start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        sums = []
        sums.append(nums[0])
        max_sum = sums[0]
        for i in range(1, len(nums)):
            sub_sum = max(sums[i - 1] + nums[i], nums[i])
            sums.append(sub_sum)
            if sums[i] > max_sum:
                max_sum = sums[i]
        return max_sum
# @lc code=end
print(Solution().maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))
